File: orchestra/limits.py
"""orchestra.limits — how much usage each account has left, and who may spend it.

Everything here hangs off one external tool: `cclimits --json`
(github.com/acrdlph/cclimits), which reads Claude Code's own usage files per
account. Shelling out is expensive and the numbers move slowly, so
`cached_limits` sits behind a 5-minute cache and only refetches from the
network when a caller explicitly asks (`refresh=True`).

The rest is policy on top of that one payload. `account_reserve` is the
buffer an account must keep free before auto-dispatch will spend it;
`_model_remaining` folds the account-wide caps (session, weekly) together
with the one model-scoped cap that the model in hand would consume;
`model_candidates` ranks the accounts that could run a mission;
`limits_by_account` is the board's per-account summary, and it is careful to
keep a maxed model-scoped cap OUT of the account-wide `exhausted` flag —
an account whose Fable is gone still has headroom for an Opus mission.
`_limit_active_until` is the resume path's question: when does the cap that
stranded this session actually reset?

`_limits` is a mutable cache dict, mutated in place and never rebound, so the
facade re-export and the tests that poke `_limits["data"]` see the same
object. Patch `limits.cached_limits`, never the facade copy.
"""

import json
import shutil
import time
from pathlib import Path

from . import config, shell


# ------------------------------------------------------------ usage limits

LIMITS_TTL_S = 300.0           # cclimits --json (its own cache) at most this often
_limits = {"t": 0.0, "data": None}


def _cclimits_bin():
    cmd = config.CFG.get("cclimits_cmd")
    if cmd:
        return cmd
    found = shutil.which("cclimits")
    if found:
        return found
    fallback = config.HOME / ".local" / "bin" / "cclimits"
    return str(fallback) if fallback.exists() else None


def cached_limits(refresh=False):
    """Per-account usage limits via cclimits (github.com/acrdlph/cclimits).
    Lazy + cached; a network refetch happens only on explicit refresh."""
    if config.DEMO:
        return demo_limits()
    now = time.time()
    if not refresh and _limits["data"] is not None and now - _limits["t"] < LIMITS_TTL_S:
        return _limits["data"]
    binp = _cclimits_bin()
    if not binp:
        return {"available": False, "error": "cclimits not found — install github.com/acrdlph/cclimits"}
    cmd = [binp, "--json"] + (["--refresh"] if refresh else [])
    rc, out = shell.run(cmd, timeout=90 if refresh else 30)
    if rc != 0 or not out:
        return _limits["data"] or {"available": False, "error": "cclimits failed (see terminal)"}
    try:
        data = json.loads(out)
    except ValueError:
        return _limits["data"] or {"available": False, "error": "cclimits returned non-JSON"}
    data["available"] = True
    data["fetched_at"] = now
    for acc in data.get("accounts", []):
        if acc.get("config_dir"):
            label = config.account_label(Path(acc["config_dir"]))
            r = account_reserve(label)
            acc["fb_label"] = label   # orchestra's label (cclimits slug may differ)
            acc["reserve_percent"] = r
            acc["reserve_blocked"] = r > 0 and (acc.get("headroom_percent") or 0) < r
    _limits["data"], _limits["t"] = data, now
    return data


def account_reserve(label):
    """Headroom % this account must keep free before auto-dispatch treats it
    as full. Per-account override, else '*' default, else 0."""
    rp = config.CFG.get("reserve_percent") or {}
    if not isinstance(rp, dict):
        return 0
    return rp.get(label, rp.get("*", 0)) or 0


def _model_remaining(acc, model):
    """Min remaining % across the limits that running `model` consumes on this
    account: all non-model-scoped limits (session, weekly) + the model-scoped
    limit matching `model`, if the account has one. None if unknown."""
    if not acc.get("ok"):
        return None
    rems = []
    for l in acc.get("limits", []):
        rem = l.get("remaining_percent")
        if rem is None:
            rem = 100 - (l.get("percent") or 0)
        if l.get("model_scoped"):
            if model and model.lower() in (l.get("label", "").lower()):
                rems.append(rem)     # this model's own cap
        else:
            rems.append(rem)         # session / weekly always apply
    return min(rems) if rems else None


def model_candidates(model, only_account=None):
    """Accounts that could run `model`, each with remaining headroom and whether
    it clears its reserve buffer. Sorted by most remaining first."""
    data = _limits["data"] if not config.DEMO else demo_limits()
    if not data or not data.get("available"):
        return []
    excl = set(config.CFG.get("exclude_accounts") or [])
    out = []
    for acc in data.get("accounts", []):
        if not acc.get("ok"):
            continue
        label = config.account_label(Path(acc["config_dir"]))
        if only_account:
            if label != only_account:
                continue
        elif label in excl:
            continue
        rem = _model_remaining(acc, model)
        if rem is None:
            continue
        reserve = account_reserve(label)
        out.append({"label": label, "remaining": round(rem),
                    "reserve": reserve, "ok": rem > 0 and rem >= reserve})
    out.sort(key=lambda x: -x["remaining"])
    return out


def set_reserve(label, percent):
    """Set an account's reserve buffer from the UI: update config.CFG, persist to the
    config file, and re-apply to the cached limits so it takes effect at once."""
    if not label:
        return {"ok": False, "error": "no account"}
    try:
        percent = max(0, min(95, int(percent)))
    except (TypeError, ValueError):
        return {"ok": False, "error": "percent must be a number"}
    rp = dict(config.CFG.get("reserve_percent") or {})
    if percent == 0:
        rp.pop(label, None)
    else:
        rp[label] = percent
    config.CFG["reserve_percent"] = rp
    # persist: merge into the on-disk config (create if missing)
    try:
        disk = {}
        if config.CONFIG_PATH and config.CONFIG_PATH.is_file():
            disk = json.loads(config.CONFIG_PATH.read_text())
        disk["reserve_percent"] = rp
        config.CONFIG_PATH.write_text(json.dumps(disk, indent=2) + "\n")
    except (OSError, ValueError) as e:
        return {"ok": False, "error": f"couldn't write config: {e}"}
    # re-enrich cached limits so the change shows without a refetch
    data = _limits.get("data")
    if data and data.get("accounts"):
        for acc in data["accounts"]:
            if acc.get("config_dir"):
                r = account_reserve(config.account_label(Path(acc["config_dir"])))
                acc["reserve_percent"] = r
                acc["reserve_blocked"] = r > 0 and (acc.get("headroom_percent") or 0) < r
    return {"ok": True, "label": label, "percent": percent}


def limits_by_account():
    """account label -> {exhausted, worst, resets_in, headroom, reserve, available,
    scoped_exhausted}.

    `exhausted`/`available` reflect only ACCOUNT-WIDE caps (session + umbrella
    weekly). A maxed model-scoped cap (e.g. Fable) is a per-model constraint,
    not an account-wide block — it lands in `scoped_exhausted` instead, so an
    account whose Fable is gone but that still has 40% all-model headroom stays
    pickable for an Opus/Sonnet mission. Collapsing every limit into one
    exhausted flag wrote such accounts off wholesale."""
    data = _limits["data"] if not config.DEMO else demo_limits()
    if not data or not data.get("available"):
        return {}
    fetched = (data.get("fetched_at") or time.time())
    out = {}
    for acc in data.get("accounts", []):
        if not acc.get("ok"):
            continue
        label = config.account_label(Path(acc["config_dir"]))
        ex = [l for l in acc.get("limits", []) if l.get("exhausted_now")]
        blocking = [l for l in ex if not l.get("model_scoped")]   # session / umbrella weekly
        worst = min(blocking, key=lambda l: l.get("resets_in_seconds") or 0) if blocking else None
        resets_in = worst.get("resets_in_seconds") if worst else None
        headroom = acc.get("headroom_percent")
        reserve = account_reserve(label)
        # reserve-blocked: less than the required buffer remains → treat as full
        reserve_blocked = reserve > 0 and headroom is not None and headroom < reserve
        out[label] = {
            "headroom": headroom,
            "exhausted": bool(blocking),
            "worst": worst["label"] if worst else None,
            "worst_scoped": False,   # `worst` is always an account-wide cap now
            "group": worst.get("group") if worst else None,
            "resets_in": resets_in,
            "resets_at": fetched + resets_in if resets_in else None,
            "reserve": reserve,
            "reserve_blocked": reserve_blocked,
            # model-scoped caps that are used up — only strand a session
            # actually running that model, not the whole account
            "scoped_exhausted": [
                {"label": l.get("label"), "group": l.get("group"),
                 "resets_in": l.get("resets_in_seconds"),
                 "resets_at": (fetched + l["resets_in_seconds"]) if l.get("resets_in_seconds") else None}
                for l in ex if l.get("model_scoped")],
            # usable for AUTO dispatch: real all-model headroom above its buffer
            "available": (not blocking) and not reserve_blocked,
        }
    return out


def demo_limits():
    now = time.time()
    def lim(label, group, pct, ex, resets_h, scoped=False):
        return {"label": label, "group": group, "percent": pct,
                "remaining_percent": 100 - pct, "model_scoped": scoped,
                "exhausted_now": ex, "resets_at": None,
                "resets_in_seconds": resets_h * 3600}
    return {"available": True, "fetched_at": now, "generated_at": None, "accounts": [
        {"slug": "default", "email": None, "plan": "max", "config_dir": "~/.claude",
         "ok": True, "error": None, "headroom_percent": 62.0, "limits": [
            lim("Session", "session", 21, False, 3.2), lim("Weekly", "weekly", 38, False, 96)]},
        {"slug": "work", "email": None, "plan": "max", "config_dir": "~/.claude-work",
         "ok": True, "error": None, "headroom_percent": 0.0, "limits": [
            lim("Session", "session", 100, True, 2.1), lim("Weekly", "weekly", 91, False, 30)]},
        {"slug": "spare", "email": None, "plan": "pro", "config_dir": "~/.claude-spare",
         "ok": True, "error": None, "headroom_percent": 88.0, "limits": [
            lim("Session", "session", 4, False, 4.8), lim("Weekly", "weekly", 12, False, 120)]},
    ]}


def _limit_active_until(account, model, now):
    """The freshest word on whether `account` still blocks this session: a
    future reset timestamp while it does, None once it's clear. Refetches
    cclimits — the cached view predates the reset by design. An unreadable
    account verifies as clear: the send costs nothing if the limit holds."""
    data = cached_limits(refresh=True)
    if not data or not data.get("available"):
        return None
    al = limits_by_account().get(account)
    if not al:
        return None
    cands = []
    if al.get("exhausted") and al.get("resets_at"):
        cands.append(al["resets_at"])         # account-wide cap bites every model
    for sx in al.get("scoped_exhausted", []):
        if not sx.get("resets_at"):
            continue
        # a model-scoped cap only blocks a session running that model; with the
        # model unknown, count it — a late resume beats a wasted one
        if not model or (sx.get("label") or "").lower() in model.lower():
            cands.append(sx["resets_at"])
    future = [c for c in cands if c > now + 30]
    return min(future) if future else None

File: ios/Sources/Orchestra/Model/Limits.swift
import Foundation

/// `GET /api/limits`, modelled from a live capture on 2026-07-22 (5,831 B, six
/// accounts).
///
/// **Two fields called `generated_at` on this API are different types.**
/// `/api/state.generated_at` is a float epoch; this one is an **ISO-8601
/// string** on the real path (it comes straight out of `cclimits`) and `null` in
/// demo mode (`limits.demo_limits`). A client that reached for the same decoder
/// for both gets `typeMismatch` and loses the screen. It is decoded as a string
/// here, and `fetched_at` — which IS a float epoch, written by orchestra itself
/// — is what the "fetched 4m ago" line uses.
public struct LimitsReport: Sendable, Equatable, Decodable {
    /// False when `cclimits` is missing or failed. Then `accounts` is empty and
    /// `error` is the whole screen — `UX.md` §3.7's whole-page error state.
    public let available: Bool
    public let error: String?
    /// orchestra's own clock, when this was fetched. The one usable timestamp.
    public let fetchedAt: Double?
    /// `cclimits`'s own stamp, ISO-8601, or absent. Display only.
    public let generatedAt: String?
    public let accounts: [AccountLimits]

    public init(available: Bool, error: String?, fetchedAt: Double?,
                generatedAt: String?, accounts: [AccountLimits]) {
        self.available = available
        self.error = error
        self.fetchedAt = fetchedAt
        self.generatedAt = generatedAt
        self.accounts = accounts
    }

    enum CodingKeys: String, CodingKey {
        case available, error, accounts
        case fetchedAt = "fetched_at"
        case generatedAt = "generated_at"
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        available = try c.decodeIfPresent(Bool.self, forKey: .available) ?? false
        error = try c.decodeIfPresent(String.self, forKey: .error)
        fetchedAt = try c.decodeIfPresent(Double.self, forKey: .fetchedAt)
        generatedAt = try c.decodeIfPresent(String.self, forKey: .generatedAt)
        accounts = try c.decodeIfPresent([AccountLimits].self, forKey: .accounts) ?? []
    }

    public var fetched: Date? { fetchedAt.map { Date(timeIntervalSince1970: $0) } }

    /// Most headroom first. The board's own auto-pick ranks the same way, so the
    /// top row is the account a dispatch would land on.
    public var ranked: [AccountLimits] {
        accounts.sorted { ($0.headroomPercent ?? -1) > ($1.headroomPercent ?? -1) }
    }
}

public struct AccountLimits: Sendable, Equatable, Decodable, Identifiable {
    /// `cclimits`'s slug — NOT orchestra's account label. The two disagree
    /// routinely (`default` here is `main` on the board), which is exactly why
    /// `fb_label` exists.
    public let slug: String
    /// **orchestra's own label**, and the only one that matches a session's
    /// `account` on the board. Absent when the account has no `config_dir`.
    public let fbLabel: String?
    public let email: String?
    public let plan: String?
    public let configDir: String?
    /// False when this one account could not be read; `error` says why.
    public let ok: Bool
    public let error: String?
    /// The minimum remaining across the account's non-model-scoped limits.
    public let headroomPercent: Double?
    public let limits: [LimitBar]
    /// The buffer auto-dispatch keeps free. Absent when there is no `config_dir`.
    public let reservePercent: Int?
    /// Headroom is below the reserve: auto-dispatch will not pick this account,
    /// **but a person still can**. Saying only the first half is the copy trap.
    public let reserveBlocked: Bool

    public var id: String { slug }

    public init(slug: String, fbLabel: String?, email: String?, plan: String?,
                configDir: String?, ok: Bool, error: String?, headroomPercent: Double?,
                limits: [LimitBar], reservePercent: Int?, reserveBlocked: Bool) {
        self.slug = slug
        self.fbLabel = fbLabel
        self.email = email
        self.plan = plan
        self.configDir = configDir
        self.ok = ok
        self.error = error
        self.headroomPercent = headroomPercent
        self.limits = limits
        self.reservePercent = reservePercent
        self.reserveBlocked = reserveBlocked
    }

    enum CodingKeys: String, CodingKey {
        case slug, email, plan, ok, error, limits
        case fbLabel = "fb_label"
        case configDir = "config_dir"
        case headroomPercent = "headroom_percent"
        case reservePercent = "reserve_percent"
        case reserveBlocked = "reserve_blocked"
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        slug = try c.decodeIfPresent(String.self, forKey: .slug) ?? "?"
        fbLabel = try c.decodeIfPresent(String.self, forKey: .fbLabel)
        email = try c.decodeIfPresent(String.self, forKey: .email)
        plan = try c.decodeIfPresent(String.self, forKey: .plan)
        configDir = try c.decodeIfPresent(String.self, forKey: .configDir)
        ok = try c.decodeIfPresent(Bool.self, forKey: .ok) ?? false
        error = try c.decodeIfPresent(String.self, forKey: .error)
        headroomPercent = try c.decodeIfPresent(Double.self, forKey: .headroomPercent)
        limits = try c.decodeIfPresent([LimitBar].self, forKey: .limits) ?? []
        reservePercent = try c.decodeIfPresent(Int.self, forKey: .reservePercent)
        reserveBlocked = try c.decodeIfPresent(Bool.self, forKey: .reserveBlocked) ?? false
    }

    /// What the board calls this account. Falls back to the `cclimits` slug so a
    /// row is never nameless.
    public var label: String { fbLabel ?? slug }

    /// Any limit that is out right now — including a model-scoped one, which
    /// does NOT block the account and must not be rendered as if it did.
    public var exhausted: [LimitBar] { limits.filter(\.exhaustedNow) }

    /// The account itself is out only when a limit that is not model-scoped is.
    public var accountExhausted: Bool {
        limits.contains { $0.exhaustedNow && !$0.modelScoped }
    }
}

/// One bar. `percent` is used, `remaining_percent` is left — the server ships
/// both and they are not always exact complements, so neither is derived here.
public struct LimitBar: Sendable, Equatable, Decodable, Identifiable {
    public let label: String
    /// `session` or `weekly`.
    public let group: String
    public let percent: Double
    public let remainingPercent: Double?
    /// A model cap. **Exhausting one blocks only sessions running that model;
    /// the account is still usable.** Collapsing that distinction is an explicit
    /// anti-goal in the server and in `UX.md` §3.7.
    public let modelScoped: Bool
    public let exhaustedNow: Bool
    /// Absolute epoch — countdowns never leave the server
    /// (`limits._absolutise_resets`). Null for a limit with no reset.
    public let resetsAt: Double?

    public var id: String { "\(group)/\(label)" }

    public init(label: String, group: String, percent: Double, remainingPercent: Double?,
                modelScoped: Bool, exhaustedNow: Bool, resetsAt: Double?) {
        self.label = label
        self.group = group
        self.percent = percent
        self.remainingPercent = remainingPercent
        self.modelScoped = modelScoped
        self.exhaustedNow = exhaustedNow
        self.resetsAt = resetsAt
    }

    enum CodingKeys: String, CodingKey {
        case label, group, percent
        case remainingPercent = "remaining_percent"
        case modelScoped = "model_scoped"
        case exhaustedNow = "exhausted_now"
        case resetsAt = "resets_at"
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        label = try c.decodeIfPresent(String.self, forKey: .label) ?? "?"
        group = try c.decodeIfPresent(String.self, forKey: .group) ?? ""
        percent = try c.decodeIfPresent(Double.self, forKey: .percent) ?? 0
        remainingPercent = try c.decodeIfPresent(Double.self, forKey: .remainingPercent)
        modelScoped = try c.decodeIfPresent(Bool.self, forKey: .modelScoped) ?? false
        exhaustedNow = try c.decodeIfPresent(Bool.self, forKey: .exhaustedNow) ?? false
        resetsAt = try c.decodeIfPresent(Double.self, forKey: .resetsAt)
    }

    public var resets: Date? { resetsAt.map { Date(timeIntervalSince1970: $0) } }
    /// Clamped for drawing only. The number printed is the server's.
    public var fraction: Double { min(1, max(0, percent / 100)) }
}
